rs/rust_canisters/random_traffic_test/src/main.rs

```diff
@@ -0,0 +1,236 @@
+    F: FnOnce(T) -> T,
+{
+    let msg = match candid::Decode!(&api::arg_data()[..], T) {
+        Ok(item) => Ok(f(item)),
+        Err(_) => Err(()),
+    };
+
+    let msg = candid::Encode!(&msg).unwrap();
+    api::reply(&msg[..]);
+}
+
+/// Replaces the test config.
+#[export_name = "canister_update replace_config"]
+fn replace_config() {
+    replace_state(|config: Config| CONFIG.replace(config));
+}
+
+/// Replaces the requests per round to be sent each heart beat.
+#[export_name = "canister_update replace_max_calls_per_heartbeat"]
+fn replace_max_calls_per_heartbeat() {
+    replace_state(|max_calls_per_heartbeat: u32| {
+        MAX_CALLS_PER_HEARTBEAT.replace(max_calls_per_heartbeat)
+    });
+}
+
+/// Replaces the reply weight.
+#[export_name = "canister_update replace_reply_weight"]
+fn replace_reply_weight() {
+    replace_state(|reply_weight: u32| REPLY_WEIGHT.replace(reply_weight));
+}
+
+/// Replaces the call weight.
+#[export_name = "canister_update replace_call_weight"]
+fn replace_call_weight() {
+    replace_state(|call_weight: u32| CALL_WEIGHT.replace(call_weight));
+}
+
+/// Seeds `RNG`.
+#[export_name = "canister_update seed_rng"]
+fn seed_rng() {
+    RNG.with_borrow_mut(|rng| {
+        let seed = candid::Decode!(&api::arg_data()[..], u64).unwrap();
+        *rng = StdRng::seed_from_u64(seed);
+    });
+    api::reply(&[]);
+}
+
+/// Returns the canister records.
+#[export_name = "canister_query records"]
+fn records() {
+    let records = RECORDS.with_borrow(|records| records.values().cloned().collect::<Vec<_>>());
+    let msg = candid::Encode!(&records).unwrap();
+    api::reply(&msg[..]);
+}
+
+/// Returns a random receiver if any.
+fn choose_receiver() -> Option<CanisterId> {
+    CONFIG.with_borrow(|config| {
+        RNG.with_borrow_mut(|rng| config.receivers.as_slice().choose(rng).cloned())
+    })
+}
+
+/// Wrapper to generate a random `u32` from one of the ranges in `Config`.
+fn gen_range<F>(f: F) -> u32
+where
+    F: FnOnce(&Config) -> RangeInclusive<u32>,
+{
+    CONFIG.with_borrow(|config| RNG.with_borrow_mut(|rng| rng.gen_range(f(config))))
+}
+
+/// Returns the next call id for use in keeping records.
+fn next_call_id() -> u32 {
+    let id = CALL_ID.take();
+    CALL_ID.set(id + 1);
+    id
+}
+
+/// Attemps to call a randomly chosen `receiver` with a random payload size. Records calls
+/// attempted in `RECORDS` in the order they were made. Once a reply is received, this record is
+/// updated in place.
+///
+/// `on_response` is executed upon awaiting the outcome of the call, both on reply and on reject.
+/// By contrast, it is important to report a synchronous rejection without calling `on_response`
+/// because we could otherwise get stuck attempting new calls indefinitely.
+fn try_call(on_response: impl FnOnce() + Copy + 'static) -> Result<(), ()> {
+    let receiver = choose_receiver().ok_or(())?;
+    let payload_bytes = gen_range(|config| config.call_bytes_min..=config.call_bytes_max);
```

Comment:
The point is to encapsulate the clumsy `with_borrow_mut()` syntax of working with `RefCell`. I don't see how these functions would help here tbh.

Reply:
You could write something like:

```suggestion
    let payload_bytes = gen_range(CONFIG.with_borrow(|config| config.call_bytes_range()));
```

You could also define them as free functions:

```rust
fn call_bytes_range() -> RangeInclusive<u32> {
    CONFIG.with_borrow(|config| config.call_bytes_min..=config.call_bytes_max)
}

fn gen_range(range: RangeInclusive<u32>) -> u32 {
    RNG.with_borrow_mut(|rng| rng.gen_range(range))
}

fn try_call(/* ... */) {
    // ...

    let payload_bytes = gen_range(call_bytes_range());

    // ...
}
```

Reply:
I can do that but I honestly don't see why
```
let payload_bytes = gen_range(|config| config.call_bytes_min..=config.call_bytes_max);
```
isn't perfectly readable as it is.

Reply:
That may well be, but this is certainly less readable than it could be (see above):

```rust
fn gen_range<F>(f: F) -> u32
where
    F: FnOnce(&Config) -> RangeInclusive<u32>,
{
```

That being said, this was very much a nit, so feel free to leave it as is.

Reply:
It is easier to read yes, but also more verbose as in extra functions. Anyways - changed it.